pkg/istio-agent/xds_proxy.go

```diff
@@ -412,35 +405,30 @@ func (p *XdsProxy) handleUpstreamResponse(con *ProxyConnection) {
 	}
 }
 
-func (p *XdsProxy) handleUpstreamECDSResponse(con *ProxyConnection) {
-	// TODO(bianpengyuan): Add metrics about ecds handling.
-	for {
-		select {
-		case resp := <-p.ecdsUpdateChan:
-			sendNack := wasm.MaybeConvertWasmExtensionConfig(resp.Resources, p.wasmCache)
-			if sendNack {
-				con.requestsChan <- &discovery.DiscoveryRequest{
-					VersionInfo:   p.ecdsLastAckVersion.Load(),
-					TypeUrl:       v3.ExtensionConfigurationType,
-					ResponseNonce: resp.Nonce,
-					ErrorDetail: &google_rpc.Status{
-						// TODO(bianpengyuan): make error message more informative.
-						Message: "failed to fetch wasm module",
-					},
-				}
-				return
-			}
-			proxyLog.Debugf("forward ECDS resources %+v", resp.Resources)
-			forwardToEnvoy(con, resp)
-
-		case <-con.stopChan:
-			return
+func (p *XdsProxy) rewriteAndForward(con *ProxyConnection, resp *discovery.DiscoveryResponse) {
+	sendNack := wasm.MaybeConvertWasmExtensionConfig(resp.Resources, p.wasmCache)
+	if sendNack {
+		proxyLog.Debugf("sending NACK for ECDS resources %+v", resp.Resources)
+		con.requestsChan <- &discovery.DiscoveryRequest{
+			VersionInfo:   p.ecdsLastAckVersion.Load(),
+			TypeUrl:       v3.ExtensionConfigurationType,
+			ResponseNonce: resp.Nonce,
+			ErrorDetail: &google_rpc.Status{
+				// TODO(bianpengyuan): make error message more informative.
+				Message: "failed to fetch wasm module",
+			},
 		}
+		return
 	}
+	proxyLog.Debugf("forward ECDS resources %+v", resp.Resources)
+	forwardToEnvoy(con, resp)
 }
 
 func forwardToEnvoy(con *ProxyConnection, resp *discovery.DiscoveryResponse) {
-	// TODO: Validate the known type urls before forwarding them to Envoy.
+	if !v3.IsEnvoyType(resp.TypeUrl) {
+		proxyLog.Errorf("Skipping forwarding type url %s to Envoy as is not a valid Envoy type", resp.TypeUrl)
```

Comment:
did you ever hit this case? or just in case?

Reply:
We did not. Just a safety net.